mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -278,26 +278,26 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
 
     private Mono<Task.Result> reIndexMessages(Flux<Either<Failure, ReIndexingEntry>> entriesToIndex, RunningOptions runningOptions, ReprocessingContext reprocessingContext) {
         return ReactorUtils.Throttler.<Either<Failure, ReIndexingEntry>, Task.Result>forOperation(
-                entry -> reIndexMessage(entry, reprocessingContext))
+                entry -> reIndex(entry, reprocessingContext))
             .window(runningOptions.getMessagesPerSecond(), Duration.ofSeconds(1))
             .throttle(entriesToIndex)
             .reduce(Task::combine)
             .switchIfEmpty(Mono.just(Result.COMPLETED));
     }
 
-    private Mono<Task.Result> reIndexMessage(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) {
-        return toMono(entryOrFailure.right().map(this::index))
-            .map(this::flapMapRight)
-            .map(either -> recordIndexingResult(reprocessingContext, either));
+    private Mono<Task.Result> reIndex(Either<Failure, ReIndexingEntry> entryOrFailure, ReprocessingContext reprocessingContext) {
+        return entryOrFailure.map(this::index)
+            .fold(failure -> {
+                failure.recordFailure(reprocessingContext);
+                return Mono.just(Result.PARTIAL);
+            }, result -> result.map(either -> recordIndexingResult(either, reprocessingContext)));
```

Comment:
As far as I understand it what you can 'result' is a `Mono<Either<Failure, ReIndexingEntry>>` while the overall type is `Either<Failure, Mono<Either<Failure, Result>>>`.

I would expect a trainsformation to go through these steps

 - Move the Mono as the main type: `Mono<Either<Failure, Either<Failure, Result>>>`
 - Then "flatmap" the right part of the expression in order to get a `Mono<Either<Failure, Result>>` wich we would be able to record failures in a single place.

(These transformation are doable as either is a monad, and would IMO lead to a cleaner code style here).

Note that method extraction helps smoothing things (exemple with the previous functionJava version).